echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy

```diff
@@ -0,0 +1,210 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.scheduler.actions.pipeline.impl
+
+import com.netflix.spinnaker.echo.model.Pipeline
+import com.netflix.spinnaker.echo.model.Trigger
+import com.netflix.spinnaker.echo.pipelinetriggers.PipelineCache
+import com.netflix.spinnaker.echo.pipelinetriggers.orca.OrcaService
+import com.netflix.spinnaker.echo.pipelinetriggers.orca.OrcaService.PipelineResponse
+import com.netflix.spinnaker.echo.pipelinetriggers.orca.PipelineInitiator
+import groovy.util.logging.Slf4j
+import org.quartz.CronExpression
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.actuate.metrics.CounterService
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression
+import org.springframework.context.ApplicationListener
+import org.springframework.context.event.ContextRefreshedEvent
+import org.springframework.stereotype.Component
+import rx.Observable
+import rx.Scheduler
+import rx.Subscription
+
+import java.time.Clock
+import java.time.Instant
+import java.time.ZoneId
+import java.time.temporal.ChronoUnit
+import java.util.concurrent.TimeUnit
+
+/**
+ * Finds and executes all pipeline triggers that should have run in the last configured time window during startup.
+ * This job will wait until the {@link com.netflix.spinnaker.echo.pipelinetriggers.PipelineCache} has run prior to
+ * finding any missed triggers.
+ */
+@ConditionalOnExpression('${scheduler.enabled:false && scheduler.compensationJob.enabled:false}')
+@Component
+@Slf4j
+class MissedPipelineTriggerCompensationJob implements ApplicationListener<ContextRefreshedEvent> {
+
+  final static int POLLING_INTERVAL_SECONDS = 5
+
+  final Scheduler scheduler
+  final PipelineCache pipelineCache
+  final OrcaService orcaService
+  final CounterService counter
+  final PipelineInitiator pipelineInitiator
+  final DateContext dateContext
+
+  transient Subscription subscription
+
+  @Autowired
+  MissedPipelineTriggerCompensationJob(Scheduler scheduler,
+                                       PipelineCache pipelineCache,
+                                       OrcaService orcaService,
+                                       PipelineInitiator pipelineInitiator,
+                                       CounterService counter,
+                                       @Value('${scheduler.compensationJob.windowMs:30000}') long compensationWindowMs,
```

Comment:
This is probably a little small for a default window?

Just to confirm the behavior ... we'll auto-kickoff a pipeline for any triggers that should have fired between `now` and `now - compensationWindowMs`?

In such scenarios we would want to err on the side of a larger window?